src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs

```diff
@@ -26,6 +26,23 @@ private CSharpSemanticFactsService()
         {
         }
 
+        protected override IEnumerable<ISymbol> GetCollidableSymbols(SemanticModel semanticModel, SyntaxNode location, SyntaxNode container, CancellationToken cancellationToken)
+        {
+            // Get all the symbols visible to the current location.
+            var visibleSymbols = semanticModel.LookupSymbols(location.SpanStart);
+
+            // Some symbols in the enclosing block could cause conflicts even if they are not available at the location.
+            // E.g. symbols inside if statements / try catch statements.
+            // Walk through the enclosing block to find them, but avoid exploring local functions because
+            //     a) Visible local function symbols would be returned from LookupSymbols
+            //        (e.g. location is inside a local function or the local function method name).
+            //     b) Symbols declared inside a local function do not cause collisions with symbols declared outside the local function, so avoid considering them.
+            var symbolsInBlock = semanticModel.GetExistingSymbols(container, cancellationToken,
+                descendInto: n => !n.IsKind(SyntaxKind.LocalFunctionStatement));
```

Comment:
How about lambda functions?

Reply:
Yeah, it looks like those don't collide either.  So we should do the same.  Note: i just realized this collision logic depends on language version.  So we need to actually check the lang version and only do this if allowed.

Reply:
> 
> 
> How about lambda functions?

THis should be handled here - http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/SemanticModelExtensions.cs,240 but I am working on writing a test for it.  Encountering issues with have a lambda next to a markup span in test files...

Reply:
> 
> 
> Yeah, it looks like those don't collide either. So we should do the same. Note: i just realized this collision logic depends on language version. So we need to actually check the lang version and only do this if allowed.

I'm don't quite get this - do you mean that local functions only apply to C# 7?  Wouldn't that automatically be handled because we'd never see the SyntaxKind.LocalFunctionStatement?

Reply:
Feel free to hit me up on gitter, but allow me to explain:

For the purposes of explanation i'm going to make up a version number (since i don't actually know when things cahnged).  So, in C# 6.0 this was **illegal**:

```c#
int i;
Action a = () => { int i; }
```

This causes a collision between the two `int i` variables that hte language explicitly disallows.  At some point (maybe 7.0, 7.1, 7.2, 7.3 or 8.0,  i honestly don't know) we decided to relax this and state that there would be no collision here.

However, this helper function is being used to say "what are the symbols i woud collide with" so that features can decide what names are available.  So, if you have this code:

```c#
$$ // <-- location of interest
Action a = () => { int i; }
```

then in C# 6.0 we need to say "`int i` would collide here, so you can't use `i`".  But in c# 7.0 we would say that "`int i` is not in scope here, so you can use `i`".

In effect, the decision we make (and what the APi returns) is language version dependent.  The rules literally changed between versions, and that impacts what features are allowed to do.

Let me know if you need more info.  Cheers!

Reply:
Ahhh, I didn't realize that the first case used to be illegal.  So yup it makes sense that we only want to exclude lambda functions based on the language version.  I'll go ahead and figure out when that happened and update the PR.

That was a perfect explanation, thanks!

Reply:
Glad to help.  Note: to make your life easier, we have an IsAnonymousOrLocalFunction helper that you can use for this purpose.

Reply:
According to my testing on VS, it would be C# 8 that relaxes those restrictions.

> 
> 
> Glad to help! Note: to make your life easier, we have an IsAnonymousOrLocalFunction helper that you can use for this purpose.

I will re-use it because I don't want to duplicate stuff, but unfortunately I think you just made my life much harder 😂
... since GenerateUniqueName only has the semantic model and syntax nodes, I'm going to have to pass in ISyntaxFactsService.  It looks like there's a nice long chain I have to update from where I can retrieve the service to where it gets passed into GenerateUniqueName